functions/src/index.ts

```diff
@@ -148,14 +170,140 @@ exports.calculateAqi = functions.pubsub
       };
 
       currentData[purpleAirId] = currentSensorData;
+
+      // Update the AQI circular buffer for this element
+      const sensorDocRef = firestore.collection('/sensors').doc(docId);
+      addToAqiBuffer(aqiBufferData, sensorDocRef);
     }
 
-    await firestore.collection('current-reading').doc('sensors').set({
+    // Send AQI reading to current-readings to be displayed on the map
+    await firestore.collection('current-reading').doc('pm25').set({
       lastUpdated: FieldValue.serverTimestamp(),
       data: currentData,
     });
   });
 
+/**
+ * This function adds an element to a sensor's AQI buffer (if applicable)
+ * @param data - the element to add to the buffer
+ * @param docRef - the document reference for a particular sensor
+ */
+function addToAqiBuffer(
+  data: AqiBufferElement,
+  docRef: FirebaseFirestore.DocumentReference<FirebaseFirestore.DocumentData>
+) {
+  docRef.get().then(doc => {
+    if (doc.exists) {
+      const status = doc.get('aqiBufferStatus');
+      if (status === bufferStatus.Exists) {
+        // The buffer exists, proceed with normal update
+        let aqiBufferIndex = doc.get('aqiBufferIndex');
+        const aqiBuffer = doc.get('aqiBuffer');
+        aqiBuffer[aqiBufferIndex] = data;
+        /* eslint-disable-next-line no-magic-numbers */
+        aqiBufferIndex = (aqiBufferIndex + 1) % aqiBuffer.length;
+
+        docRef.update({
+          aqiBufferIndex: aqiBufferIndex,
+          aqiBuffer: aqiBuffer,
+        });
+      } else if (status === bufferStatus.DoesNotExist || status === undefined) {
+        // Populate the buffer with default values, skip this sensor
+        docRef.update({
+          aqiBufferStatus: bufferStatus.InProgress,
+        });
+        populateDefaultBuffer(true, docRef);
+      }
+    }
+  });
+}
+
+/**
+ * This function adds an element to a sensor's PM2.5 buffer (if applicable)
+ * @param data - the element to add to the buffer
+ * @param docRef - the document reference for a particular sensor
+ */
+function addToPm25Buffer(
```

Comment:
This code is similar to the last function. I think it's easier to read without being littered with if statements, but there is a lot of duplicate so if anyone else prefers that these become one function with if statements instead, I can do that.